OCPBUGS-69447: feat(updates): enable CVO metrics access with RHOBS monitoring flag

Chee-Lu opened the pull request:
## What this PR does:
When `--rhobs-monitoring=true` is set (for ROSA HCP), enable CVO access to OBO Prometheus for conditional update risk evaluation. 

Aldo add `--cvo-prometheus-url` flag to allow overriding the default Prometheus endpoint. This provides flexibility for future changes (e.g., if ROSA changes the service name) or for platforms with different monitoring architectures (e.g., ARO HCP's self-managed Prometheus). When not specified, platform-appropriate defaults are used.

The CVO deployment logic routes to different metrics endpoints based on the monitoring stack:
- RHOBS stack (ROSA HCP): `http://hypershift-monitoring-stack-prometheus.openshift-observability-operator.svc:9090`
- CoreOS stack (Self-managed HyperShift on OpenShift or ARP HCP): `https://thanos-querier.openshift-monitoring.svc:9092`
- Custom URL: Configurable via `--cvo-prometheus-url` parameter

Key changes:
- CVO deployment enables metrics access when either `--rhobs-monitoring` (for ROSA HCP) or `--enable-cvo-management-cluster-metrics-access` (for self-managed HyperShift on OpenShift or ARO HCP) is set
- Network policies updated to allow egress to the appropriate monitoring endpoint based on stack configuration
- Add `--cvo-prometheus-url` flag to configure CVO Prometheus endpoint

## Which issue(s) this PR fixes:
fixes https://issues.redhat.com//browse/OCM-10395
fixes https://issues.redhat.com//browse/OCM-20970

## Special notes for your reviewer:
### Backport Requirements
This change should be backported to 4.17.z, 4.18.z, 4.19.z, 4.20.z and 4.21.z to benefit customers on that version.

sdminonne commented:
@coderabbitai review

Chee-Lu commented on cmd/install/install.go:
```diff
@@ -269,7 +269,7 @@ func NewCommand() *cobra.Command {
 	cmd.Flags().BoolVar(&opts.WaitUntilAvailable, "wait-until-available", opts.WaitUntilAvailable, "If true, pauses installation until hypershift operator has been rolled out and its webhook service is available (if installing the webhook)")
 	cmd.Flags().BoolVar(&opts.WaitUntilEstablished, "wait-until-crds-established", opts.WaitUntilEstablished, "If true, pauses installation until all custom resource definitions are established before applying other manifests.")
-	cmd.PersistentFlags().BoolVar(&opts.RHOBSMonitoring, "rhobs-monitoring", opts.RHOBSMonitoring, "If true, HyperShift will generate and use the RHOBS version of monitoring resources (ServiceMonitors, PodMonitors, etc)")
+	cmd.PersistentFlags().BoolVar(&opts.RHOBSMonitoring, "rhobs-monitoring", opts.RHOBSMonitoring, "If true, HyperShift will generate and use the RHOBS version of monitoring resources (ServiceMonitors, PodMonitors, etc) and enable CVO access to OBO Prometheus for conditional update risk evaluation")
```
Thanks for the feedback, I've replace internal code names with public terminology

LiangquanLi930 commented on control-plane-operator/controllers/hostedcontrolplane/v2/cvo/deployment.go:
```diff
@@ -93,10 +98,24 @@ func (cvo *clusterVersionOperator) adaptDeployment(cpContext component.WorkloadC
-		if cvo.enableCVOManagementClusterMetricsAccess {
+
+		// Enable CVO metrics access if either RHOBS monitoring is enabled or the explicit flag is set
+		enableMetricsAccess := os.Getenv(rhobsmonitoring.EnvironmentVariable) == "1" || cvo.enableCVOManagementClusterMetricsAccess
```
This code has already been defined on line 33.

Chee-Lu replied:
Removed

LiangquanLi930 commented on control-plane-operator/controllers/hostedcontrolplane/v2/cvo/deployment.go:
```diff
@@ -93,10 +98,24 @@ func (cvo *clusterVersionOperator) adaptDeployment(cpContext component.WorkloadC
+			var metricsURL string
+			if os.Getenv(rhobsmonitoring.EnvironmentVariable) == "1" {
+				// RHOBS Prometheus uses HTTP without TLS
+				metricsURL = fmt.Sprintf("http://hypershift-monitoring-stack-prometheus.openshift-observability-operator.svc:9090?namespace=%s", cpContext.HCP.Namespace)
```
We can determine this using a check like the following:
```
func isROSAHCP(hc *hyperv1.HostedCluster) bool {
	if hc.Spec.Platform.AWS == nil {
		return false
	}

	for _, tag := range hc.Spec.Platform.AWS.ResourceTags {
		if tag.Key == "red-hat-managed" && tag.Value == "true" {
			return true
		}
	}
	return false
}

```

simonpasquier commented on control-plane-operator/controllers/hostedcontrolplane/v2/cvo/deployment.go:
```diff
@@ -93,10 +98,24 @@ func (cvo *clusterVersionOperator) adaptDeployment(cpContext component.WorkloadC
+			var metricsURL string
+			if os.Getenv(rhobsmonitoring.EnvironmentVariable) == "1" {
+				// RHOBS Prometheus uses HTTP without TLS
+				metricsURL = fmt.Sprintf("http://hypershift-monitoring-stack-prometheus.openshift-observability-operator.svc:9090?namespace=%s", cpContext.HCP.Namespace)
```
And about the URL, my gut feeling is that it isn't an HyperShift concern but rather something that would be provided via configuration (e.g what if ROSA decides to change the service name?).

Chee-Lu commented on hypershift-operator/controllers/hostedcluster/network_policies.go:
```diff
@@ -33,6 +35,19 @@ const (
+		}
+	}
+	return false
+}
```
fixed, moved IsROSAHCP to shared awsutil package

sjenning commented:
/approve

waiting to apply ok-to-test until CI is working again

muraee commented:
please squash the commits.
happy to tag afterwards.

muraee commented on control-plane-operator/controllers/hostedcontrolplane/v2/cvo/deployment.go:
```diff
@@ -93,10 +99,22 @@ func (cvo *clusterVersionOperator) adaptDeployment(cpContext component.WorkloadC
+			if os.Getenv(rhobsmonitoring.EnvironmentVariable) == "1" && awsutil.IsROSAHCP(cpContext.HCP) {
+				// RHOBS Prometheus uses HTTP without TLS (ROSA HCP specific)
+				c.Args = append(c.Args, "--metrics-ca-bundle-file=")
+				metricsURL = "http://hypershift-monitoring-stack-prometheus.openshift-observability-operator.svc:9090"
```
missed that comment. Agreed, I think it makes sense to make this configurable rather than hardcoding. ARO/GCP can then use the same knob if needed to configure the correct url.

Chee-Lu replied:
@simonpasquier Thanks for clarifying the monitoring stack differences across platforms!

You're right that tying the CVO URL to RHOBS API group usage wouldn't work for ARO HCP. Based on your input, I've decoupled the CVO URL configuration from RHOBS monitoring:

- Renamed `--rhobs-prometheus-url` to `--cvo-prometheus-url`
- The `--cvo-prometheus-url` parameter now works independently of the `--rhobs-monitoring` flag

This way, ARO HCP (using `monitoring.coreos.com` API group) can override the CVO URL via the `--cvo-prometheus-url` parameter with requiring `RHOBSMonitoring=true`.

Chee-Lu commented:
> Just to recap, we now have the following environment variables which are somehow related:
> 
> * `RHOBS_MONITORING`
> * `ENABLE_CVO_MANAGEMENT_CLUSTER_METRICS_ACCESS`
> * `CVO_PROMETHEUS_URL`
> 
> ENABLE_CVO_MANAGEMENT_CLUSTER_METRICS_ACCESS=0	ENABLE_CVO_MANAGEMENT_CLUSTER_METRICS_ACCESS=1
> CVO_PROMETHEUS_URL not defined	CVO doesn't run PromQL risk evals except if rhobs-monitoring is on in which case it uses the default ROSA URL	CVO runs PromQL risk evals against the OCP Thanos querier (if not rhobs-monitoring) or the default ROSA Prometheus URL
> CVO_PROMETHEUS_URL defined	same as above	CVO runs PromQL risk evals against the configured URL
> Should we prevent some combinations to improve the user experience? For instance, setting `CVO_PROMETHEUS_URL` while not setting `ENABLE_CVO_MANAGEMENT_CLUSTER_METRICS_ACCESS` doesn't make lots of sense? Or should we consider that setting `CVO_PROMETHEUS_UR`L automatically enables CVO's metrics access?
> 
> Also the impact of the "rhobs-monitoring" setting on the final CVO configuration isn't very obvious IMHO. I'd rather have the rhobs-monitoring setting influence only which API group to use for monitoring resources

hi @simonpasquier , Thanks for raising these points. I think this touches on some historical design decisions that are worth discussing.

> Also the impact of the "rhobs-monitoring" setting on the final CVO configuration isn't very obvious IMHO. 
1. Yes, from the name of flag, the impact of --rhobs-monitoring on CVO configuration isn't immediately obvious. This stems from a limitation: --rhobs-monitoring and --enable-cvo-management-cluster-metrics-access were mutually exclusive because CVO accessing RHOBS metrics was not supported at that time. Additionally, these two flags had semantically conflicting names.

I see three possible approaches to address this:

**Option 1: Rename for Separation of Concerns**

Rename --enable-cvo-management-cluster-metrics-access → --enable-cvo-metrics-access
--rhobs-monitoring: Only controls monitoring stack API group
--enable-cvo-metrics-access: Controls CVO metrics access for both CoreOS and RHOBS
Pros: Clear separation, no semantic conflict
Cons: Breaking change, requires deprecation period and backward compatibility handling

**Option 2: Add another New Flag --enable-rhobs-cvo-metrics-access (No Breaking Changes)**

--rhobs-monitoring: Only controls monitoring stack API group
Keep existing --enable-cvo-management-cluster-metrics-access for CoreOS/self-managed
Add new --enable-rhobs-cvo-metrics-access for RHOBS-specific CVO configuration
Pros: No breaking changes, explicit control
Cons: Adds complexity, ROSA users need to use --rhobs-monitoring + --enable-rhobs-cvo-metrics-access

**Option 3: Current Implementation**

This is also based on @wking 's feedback - he preferred keeping it simple with a single --rhobs-monitoring flag
--rhobs-monitoring: Controls both monitoring API and implicitly enables CVO metrics access for ROSA HCP
Keep existing --enable-cvo-management-cluster-metrics-access for self-managed HyperShift
Pros: Simple for users (one flag does everything for RHOBS), no breaking changes
Cons: The impact on CVO configuration isn't very obvious from the flag name, but we can add descriptions 

IMO, this is acceptable, So I've added back the description to --rhobs-monitoring explicitly mention that it enables CVO metrics access.


> Should we prevent some combinations to improve the user experience?
2. Yes, I've added validation in this PR: setting --cvo-prometheus-url without either --rhobs-monitoring or --enable-cvo-management-cluster-metrics-access now produces a clear error message.

> Should we consider that setting CVO_PROMETHEUS_URL automatically enables CVO's metrics access?
3. I don't think this would work well. Like if a user sets a RHOBS URL but hasn't enabled RHOBS (still using CoreOS), there would be no data available and it would fail. It's better to require explicit enablement.

Let me raise this in our meeting today.

simonpasquier commented on control-plane-operator/controllers/hostedcontrolplane/v2/cvo/deployment.go:
```diff
@@ -93,10 +109,29 @@ func (cvo *clusterVersionOperator) adaptDeployment(cpContext component.WorkloadC
+					cvoPrometheusURL = "http://hypershift-monitoring-stack-prometheus.openshift-observability-operator.svc:9090"
+				}
+				c.Args = append(c.Args, "--metrics-ca-bundle-file=/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt")
+				c.Args = append(c.Args, "--metrics-token-file=/var/run/secrets/kubernetes.io/serviceaccount/token")
```
(nit) isn't it the same as the default value used by CVO?

Chee-Lu commented on hypershift-operator/controllers/hostedcluster/network_policies.go:
```diff
@@ -831,36 +836,72 @@ func kasEndpointsToCIDRs(kubernetesEndpoint *corev1.Endpoints) []string {
+// It allows egress traffic to the HCP metrics server that is available for self-managed HyperShift and ROSA HCP.
+// Depending on the monitoring stack:
+// - RHOBS (ROSA HCP only): targets OBO Prometheus at port 9090 in openshift-observability-operator namespace
+// - CoreOS (self-managed HyperShift): targets Thanos Querier at port 9092 in openshift-monitoring namespace
```
I think it makes sense to extend this when the requirements from ARO or GCP become clearer. What do you think?

Chee-Lu commented:
/retest

Chee-Lu commented:
/retest

Chee-Lu commented:
/retest

tdrozdowski commented:
> Hi @celebdor @csrwng, all tests for this PR have passed! Could you PTAL if we need an ok-to-test or an approval before it can be merged?

@celebdor / @csrwng  - just a bump on this after the weekend.  thx.
